src/protocol/mod.rs

```diff
@@ -0,0 +1,144 @@
+    Request = 0x00,
+    Info = 0x01,
+    Data = 0x02,
+    Error = 0x03,
+}
+
+#[derive(FromPrimitive, ToPrimitive)]
+enum RequestCodes {
+    LogShow = 0x00,
+    LogAdd = 0x01,
+    LogDelete = 0x02,
+    LogList = 0x03,
+    MessageAdd = 0x04,
+    IteratorAdd = 0x05,
+    IteratorList = 0x06,
+    IteratorNext = 0x07,
+    IteratorDelete = 0x08,
+}
+
+pub struct Connection {
+    framer: Framed<TcpStream, LengthDelimitedCodec>,
+}
+
+impl Connection {
+    pub async fn next_request(&mut self) -> Option<Result<Command, Error>> {
+        let frame = match self.framer.next().await {
+            Some(f) => f,
+            None => return None,
+        };
+
+        let result = match frame {
+            Ok(bytes) => read_command(bytes),
+            Err(_) => Err(Error::FailedToReadBytes),
+        };
+
+        Some(result)
+    }
+
+    pub async fn respond(&mut self, resp: Response) {
+        debug!("responding with: {:?}", &resp);
+        if let Err(e) = self.framer.send(resp.into()).await {
+            error!("could not respond: {}", e);
+        };
+    }
+}
+
+impl From<TcpStream> for Connection {
+    fn from(socket: TcpStream) -> Connection {
+        let framer = Framed::new(socket, LengthDelimitedCodec::new());
+        Connection { framer }
+    }
+}
+
+macro_rules! parse_cbor {
+    ($cmd:ident, $data:expr) => {{
+        let c = match serde_cbor::from_slice($data) {
+            Ok(c) => c,
+            Err(_) => return Err(Error::CouldNotReadPayload),
+        };
+        Command::$cmd(c)
+    }};
+}
+
+fn read_command(bytes: BytesMut) -> Result<Command, Error> {
+    let kind = match FrameKind::from_u8(bytes[0]) {
+        Some(k) => k,
+        None => return Err(Error::UnknownFrameKind),
+    };
+
+    let code = match RequestCodes::from_u8(bytes[1]) {
+        Some(c) => c,
+        None => return Err(Error::UnknownRequestCode),
+    };
+
+    if kind != FrameKind::Request {
+        return Err(Error::ServerOnlyAcceptsRequests);
+    }
+
+    let data = &bytes[2..];
+
+    use RequestCodes::*;
+    let cmd = match code {
+        LogShow => parse_cbor!(LogShow, data),
+        LogAdd => parse_cbor!(LogAdd, data),
+        LogDelete => parse_cbor!(LogDelete, data),
+        LogList => Command::LogList,
+        MessageAdd => parse_cbor!(MessageAdd, data),
+        IteratorAdd => parse_cbor!(IteratorAdd, data),
+        IteratorList => parse_cbor!(IteratorList, data),
+        IteratorNext => parse_cbor!(IteratorNext, data),
+        IteratorDelete => parse_cbor!(IteratorDelete, data),
+    };
+
+    Ok(cmd)
+}
+
+#[derive(Debug)]
+pub enum Response {
+    Info(Vec<u8>),
+    Data(Vec<Vec<u8>>),
+    Error(Error),
+}
+
+impl From<Error> for Response {
+    fn from(e: Error) -> Response {
+        Response::Error(e)
+    }
+}
+
+impl From<Response> for Bytes {
+    fn from(r: Response) -> Bytes {
+        match r {
+            Response::Info(data) => [&[FrameKind::Info.to_u8().unwrap(), 0x00 as u8], &*data]
```

Comment:
why do you need to do these `as u8`?

Reply:
`data` is of type `Vec<u8>` which we deref into `&[u8]`, so the `&[_]` we're concating with has to have the same type.  I assume I ran into a compiler that told me to do it, but I don't know why `0x00` wouldn't just be type inferenced into a u8.